apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,520 @@
+use tracing::info;
+
+use crate::configuration::shared::Client;
+use crate::services::http::HttpClientService;
+use crate::services::http::HttpRequest;
+use crate::services::router::body::RouterBody;
+
+/// An error from the snapshot server
+#[derive(Debug, thiserror::Error)]
+enum SnapshotError {
+    /// Unable to load snapshots
+    #[error("unable to load snapshots")]
+    IoError(#[from] std::io::Error),
+    /// Unable to parse snapshots
+    #[error("unable to parse snapshots")]
+    ParseError(#[from] serde_json::Error),
+}
+
+/// A server that mocks an API using snapshots recorded from actual HTTP responses.
+#[cfg_attr(test, allow(unreachable_pub))]
+pub struct SnapshotServer {
+    // The socket address the server is listening on
+    #[cfg_attr(not(test), allow(dead_code))]
+    socket_address: SocketAddr,
+}
+
+#[derive(Clone)]
+struct SnapshotServerState {
+    client: HttpClientService,
+    base_url: Uri,
+    snapshots: Arc<Mutex<Vec<Snapshot>>>,
+    snapshot_file: Box<Path>,
+    offline: bool,
+    update: bool,
+    include_headers: Option<Vec<String>>,
+}
+
+async fn root_handler(
+    State(state): State<SnapshotServerState>,
+    req: http::Request<axum::body::Body>,
+) -> Result<http::Response<RouterBody>, StatusCode> {
+    handle(State(state), req, "/".to_string()).await
+}
+
+async fn handler(
+    State(state): State<SnapshotServerState>,
+    AxumPath(path): AxumPath<String>,
+    req: http::Request<axum::body::Body>,
+) -> Result<http::Response<RouterBody>, StatusCode> {
+    handle(State(state), req, path).await
+}
+
+async fn handle(
+    State(state): State<SnapshotServerState>,
+    req: http::Request<axum::body::Body>,
+    path: String,
+) -> Result<http::Response<RouterBody>, StatusCode> {
+    let uri = [state.base_url.to_string(), path.clone()].concat();
+    let method = req.method().clone();
+    let version = req.version();
+    let request_headers = req.headers().clone();
+    let hyper_body = hyper::body::to_bytes(req.into_body()).await.unwrap();
+    let router_body: RouterBody = RouterBody::from(hyper_body.clone());
+    let request_json_body = serde_json::from_slice(&hyper_body).unwrap_or(Value::Null);
+
+    {
+        // check if we have an existing snapshot for this request
+        let mut snapshots = state.snapshots.lock().unwrap();
+        let existing = snapshots.iter().enumerate().find(|(_, snapshot)| {
+            snapshot.request.path == Some(path.clone())
+                && snapshot.request.method == Some(method.to_string())
+                && snapshot.request.body.clone() == request_json_body
+        });
+        if let Some((i, snapshot)) = existing {
+            if state.update {
+                snapshots.remove(i);
+            } else {
+                info!(
+                    url = %uri,
+                    method = %method,
+                    "Found existing snapshot"
+                );
+                return Ok(snapshot.clone().try_into().unwrap());
```

Comment:
Since snapshots can be edited by hand is it OK to unwrap here?

Reply:
Added an error here.